ArcGISRuntimeSDKQt_CppSamples/Layers/ListKmlContents/ListKmlContents.cpp

```diff
@@ -0,0 +1,313 @@
+    }
+  });
+}
+
+QString ListKmlContents::getKmlNodeType(KmlNode *node)
+{
+  if (node == nullptr)
+    return "";
+
+  QString type = "";
+  switch (node->kmlNodeType()) {
+  case KmlNodeType::KmlDocument:
+    type = "KmlDocument";
+    break;
+  case KmlNodeType::KmlFolder:
+    type = "KmlFolder";
+    break;
+  case KmlNodeType::KmlNetworkLink:
+    type = "KmlNetworkLink";
+    break;
+  case KmlNodeType::KmlPlacemark:
+    type = "KmlPlacemark";
+    break;
+  case KmlNodeType::KmlPhotoOverlay:
+    type = "KmlPhotoOverlay";
+    break;
+  case KmlNodeType::KmlGroundOverlay:
+    type = "KmlGroundOverlay";
+    break;
+  case KmlNodeType::KmlScreenOverlay:
+    type = "KmlScreenOverlay";
+    break;
+  case KmlNodeType::KmlTour:
+    type = "KmlTour";
+    break;
+  default:
+    return "";
+  }
+  return type;
+}
+
+void ListKmlContents::displayChildren(KmlNode* parentNode)
+{
+  if (parentNode == nullptr)
+    return;
+
+  if (KmlContainer* container = dynamic_cast<KmlContainer*>(parentNode))
+  {
+    m_levelNodeNames.clear();
+    bool lastLevel = true;
+
+    // for current level, get names of child nodes
+    for (KmlNode* node: *(container->childNodesListModel()))
+    {
+      QString str = node->name() + " - " + getKmlNodeType(node);
+
+      // if node has children, add ">" to indicate further levels
+      if (!node->children().isEmpty())
+      {
+        str.append(" >");
+        lastLevel = false;
+      }
+      m_levelNodeNames << str;
+    }
+
+    if (lastLevel)
+    {
+      m_currentNode = container->childNodesListModel()->at(0);
+      emit currentNodeChanged();
+    }
+    emit levelNodeNamesChanged();
+  }
+}
+
+// recursively build string to indicate node's ancestors
+void ListKmlContents::buildPathLabel(KmlNode* node)
+{
+  if (node == nullptr)
+    return;
+
+  if (node->parentNode() != nullptr)
+  {
+    buildPathLabel(node->parentNode());
+    m_labelText.append(">");
+  }
+  m_labelText.append(node->name());
+}
+
+void ListKmlContents::displayPreviousLevel()
+{
+  KmlNode* parentNode = m_currentNode->parentNode();
+  if (parentNode == nullptr)
+    return;
+
+  KmlNode* grandparentNode = parentNode->parentNode();
+
+  if (grandparentNode != nullptr)
+  {
+    displayChildren(grandparentNode);
+    m_currentNode = grandparentNode;
+    emit currentNodeChanged();
+    emit labelTextChanged();
+  }
+  // if grandparent node is nullptr, then at top of tree
+  else
+  {
+    displayChildren(parentNode);
+    m_currentNode = parentNode;
+    emit currentNodeChanged();
+    emit labelTextChanged();
+  }
+}
+
+// display selected node on sceneview and show its children
+void ListKmlContents::processSelectedNode(const QString& nodeName)
+{
+  QString extractedNodeName = nodeName;
+  if (nodeName.contains(" - "))
+  {
+    int ind = nodeName.lastIndexOf(" - ");
+    extractedNodeName = nodeName.left(ind);
+  }
+
+  // find node in the list
+  for (KmlNode* node : m_kmlNodesList)
```

Comment:
I don't think that `m_kmlNodesList` needs to exist. We just need to loop through the immediate children of `m_curretNode`.

Reply:
It is possible to click on different nodes on the same level, i.e., siblings.
![image](https://user-images.githubusercontent.com/60744198/84142915-8e9bcb80-aa4d-11ea-9fa6-e9f62488ccc6.png)

For example in the image above, I could click on "Simple placemark" and `m_currentNode` would correspond to the "Simple placemark" node. I should then be able to click on "Floating placemark" but if I just searched through the children of `m_currentNode` I wouldn't find the "Floating placemark" node.

Reply:
So the list is a combination of current children and siblings?

Reply:
The list is built once using `buildTree` and contains all the nodes in the tree